devito/distributed.py

```diff
@@ -0,0 +1,115 @@
+        self._glb_shape = shape
+        self._dimensions = dimensions
+        self._input_comm = (input_comm or MPI.COMM_WORLD).Clone()
+
+        # `Compute_dims` sets the dimension sizes to be as close to each other
+        # as possible, using an appropriate divisibility algorithm. Thus, in 3D:
+        # * topology[0] >= topology[1] >= topology[2]
+        # * topology[0] * topology[1] * topology[2] == self._input_comm.size
+        topology = MPI.Compute_dims(self._input_comm.size, len(shape))
+        # At this point MPI's dimension 0 corresponds to the rightmost element
+        # in `topology`. This is in reverse to `shape`'s ordering. Hence, we
+        # now restore consistency
+        self._topology = tuple(reversed(topology))
+
+        if self._input_comm is not input_comm:
+            # By default, Devito arranges processes into a cartesian topology.
+            # MPI works with numbered dimensions and follows the C row-major
+            # numbering of the ranks, i.e. in a 2x3 Cartesian topology (0,0)
+            # maps to rank 0, (0,1) maps to rank 1, (0,2) maps to rank 2, (1,0)
+            # maps to rank 3, and so on.
+            self._comm = self._input_comm.Create_cart(self._topology)
+        else:
+            self._comm = input_comm
+
+        # Perform domain decomposition
+        self._glb_numbs = [np.array_split(range(i), j)
+                           for i, j in zip(shape, self._topology)]
+
+    def __del__(self):
+        self._input_comm.Free()
+
+    @property
+    def myrank(self):
+        return self._comm.rank
+
+    @property
+    def mycoords(self):
+        return tuple(self._comm.coords)
+
+    @property
+    def comm(self):
+        return self._comm
+
+    @property
+    def nprocs(self):
+        return self._comm.size
+
+    @property
+    def ndim(self):
+        return len(self._glb_shape)
+
+    @property
+    def topology(self):
+        return self._topology
+
+    @property
+    def glb_numb(self):
+        """Return the global numbering of this process' domain."""
+        assert len(self._comm.coords) == len(self._glb_numbs)
+        return tuple(i[j] for i, j in zip(self._glb_numbs, self._comm.coords))
+
+    @property
+    def shape(self):
+        """Return the shape of this process' domain."""
+        return tuple(len(i) for i in self.glb_numb)
+
+    @property
+    def dimensions(self):
+        return self._dimensions
+
+    @property
+    def neighbours(self):
+        """
+        Return the mapper ``proc -> direction``; ``proc`` is the rank of a
+        neighboring process, while ``direction`` tells whether ``proc`` is
+        logically at right (value=1) or left (value=-1) of ``self``.
+        """
+        shifts = {d: self._comm.Shift(i, 1) for i, d in enumerate(self.dimensions)}
+        ret = {}
+        for d, (src, dest) in shifts.items():
+            ret[d] = {}
+            ret[d][LEFT] = src if src != -1 else None
```

Comment:
`src>-1`, can be lot more negative
All test pass with this change

Reply:
already dropped this, as you'll see in the following PR